### ISSUE TITLE
Model perform well when using flash_attention_2 or SDPA, but it's performance drops when using the original attention (i.e., attn_implementation="eager")

### ISSUE THREAD

**Issue by WellDonePF** [LLaVA-VL/LLaVA-NeXT#254]
I found this issue when working with the lmms-lab/llava-onevision-qwen2-7b-ov model and qwen2vl.（the transformers library is the latest version.）

by the way, is there a way to extract the attention score matrix (the matrix after applying softmax) when using flash attention 2 or SDPA?

### Code
```python
import json
import argparse
from PIL import Image
import copy
import torch
import accelerate

from llava.model.builder import load_pretrained_model
from llava.mm_utils import process_images, tokenizer_image_token
from llava.constants import IMAGE_TOKEN_INDEX, DEFAULT_IMAGE_TOKEN
from llava.conversation import conv_templates

def main(args):
    accelerator = accelerate.Accelerator()
    llava_model_args = {
            "multimodal": True,
        }
    overwrite_config = {}
    overwrite_config["image_aspect_ratio"] = "pad"
    llava_model_args["overwrite_config"] = overwrite_config

    tokenizer, model, image_processor, max_length = load_pretrained_model(
        model_path=args.pretrained_model_name_or_path,
        model_base=None,
        model_name=args.model_name,
        device_map=accelerator.device,
        attn_implementation=args.attn_implementation,
        **llava_model_args
    )
    model.eval()

    info = json.loads(open(args.data_path,'r',encoding="utf-8").read())
    model = accelerator.prepare(model)
    if accelerator.num_processes > 1:
        model = model.module

    cnt = -1
    for item in info:
        cnt += 1
        context = item["context"]
        context = context.replace("<ImageHere>", DEFAULT_IMAGE_TOKEN)
        images = item["raw_img_list"]
        images = [Image.open(img) for img in images]
        images = process_images(images, image_processor, model.config)
        images = [img.to(dtype=torch.float16, device=accelerator.device) for img in images]

        conv = copy.deepcopy(conv_templates[args.conv_template])
        conv.append_message(conv.roles[0], context)
        conv.append_message(conv.roles[1], None)
        prompt_question = conv.get_prompt()

        input_ids = tokenizer_image_token(prompt_question, tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt").unsqueeze(0).to(accelerator.device)
        image_sizes = [img.size for img in images]

        try:
            cont = model.generate(
                input_ids,
                images=images,
                image_sizes=image_sizes,
                do_sample=False,
                temperature=0,
                max_new_tokens=4096,
            )
        except Exception as e:
            print(e)
            continue
        result = tokenizer.batch_decode(cont, skip_special_tokens=True)
        result = result[0]

        print(result)


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--pretrained_model_name_or_path", type=str, default="lmms-lab/llava-onevision-qwen2-7b-ov")
    parser.add_argument("--model_name", type=str, default="llava_qwen")
    parser.add_argument("--root_path", type=str, default="******")
    parser.add_argument("--data_path", type=str, default="******")
    parser.add_argument("--attn_implementation", type=str, default="eager", choices=["eager", "flash_attention_2", "sdpa"])
    parser.add_argument("--conv_template", type=str, default="qwen_1_5")

    return parser.parse_args()


if __name__ == "__main__":
    main(parse_args())
```

**Comment by yfChang-cv** [comment]
same question.
When using `attn_implementation="eager"`, the “llava-onevision-qwen2-0.5b-si” model works normally, but the “llava-onevision-qwen2-7b-ov” model constantly outputs “!!!”.

**Comment by yfChang-cv** [comment]
> same question. When using `attn_implementation="eager"`, the “llava-onevision-qwen2-0.5b-si” model works normally, but the “llava-onevision-qwen2-7b-ov” model constantly outputs “!!!”.

llava-onevision-qwen2-7b-ov+sdpa:
The video features a person standing on a stage, dressed in a black shirt and dark pants. A large hand appears from the background, reaching towards the person's pocket. The text 'Source: Joshua AG' is displayed at the top left corner of the frames, and 'EVAN CARMICHAEL' is shown in the top right corner. The text 'Anyone know what this pocket is for?' appears as the hand continues to reach into the pocket. The person then looks down at their pocket, and the text 'I've always wondered that' appears. The hand finally pulls out a small white device labeled 'iPod Nano'. The person holds up the iPod Nano, and the text 'is the new iPod Nano' appears. The video concludes with a close-up of the person holding the iPod Nano, showing it from different angles.

llava-onevision-qwen2-7b-ov+eager:
The!! video!!!!!!!!!!!


llava-onevision-qwen2-0.5b-ov+sdpa:
The video features a person standing on a stage, holding a pair of jeans in front of them. The background is dark with a large hand reaching towards the jeans, causing it to appear as if someone is pulling or pushing it. The text 'Anyone know what this pocket' appears at the bottom left corner, and 'is for?' appears at the bottom right corner. The scene then transitions to show the same person now holding an iPod Nano o, which is white with a black screen. The text 'this is the new' appears at the bottom center of the frame. Finally, the person holds up a pen, indicating they are about to write something.


llava-onevision-qwen2-0.5b-ov+eager:
The video features a person standing on a stage, holding a pair of jeans in front of them. The background is dark with a large hand reaching towards the jeans, causing it to appear as if someone is pulling or pushing it. The text 'Anyone know what this pocket' appears at the bottom left corner, and 'is for?' appears at the bottom right corner. The scene then transitions to show the same person now holding an iPod Nano o, which is white with a black screen. The text 'this is the new' appears at the bottom center of the frame. Finally, the person holds up a pen, indicating they are about to write something.

**Comment by WellDonePF** [comment]
> > same question. When using `attn_implementation="eager"`, the “llava-onevision-qwen2-0.5b-si” model works normally, but the “llava-onevision-qwen2-7b-ov” model constantly outputs “!!!”.
> 
> llava-onevision-qwen2-7b-ov+sdpa: The video features a person standing on a stage, dressed in a black shirt and dark pants. A large hand appears from the background, reaching towards the person's pocket. The text 'Source: Joshua AG' is displayed at the top left corner of the frames, and 'EVAN CARMICHAEL' is shown in the top right corner. The text 'Anyone know what this pocket is for?' appears as the hand continues to reach into the pocket. The person then looks down at their pocket, and the text 'I've always wondered that' appears. The hand finally pulls out a small white device labeled 'iPod Nano'. The person holds up the iPod Nano, and the text 'is the new iPod Nano' appears. The video concludes with a close-up of the person holding the iPod Nano, showing it from different angles.
> 
> llava-onevision-qwen2-7b-ov+eager: The!! video!!!!!!!!!!!
> 
> llava-onevision-qwen2-0.5b-ov+sdpa: The video features a person standing on a stage, holding a pair of jeans in front of them. The background is dark with a large hand reaching towards the jeans, causing it to appear as if someone is pulling or pushing it. The text 'Anyone know what this pocket' appears at the bottom left corner, and 'is for?' appears at the bottom right corner. The scene then transitions to show the same person now holding an iPod Nano o, which is white with a black screen. The text 'this is the new' appears at the bottom center of the frame. Finally, the person holds up a pen, indicating they are about to write something.
> 
> llava-onevision-qwen2-0.5b-ov+eager: The video features a person standing on a stage, holding a pair of jeans in front of them. The background is dark with a large hand reaching towards the jeans, causing it to appear as if someone is pulling or pushing it. The text 'Anyone know what this pocket' appears at the bottom left corner, and 'is for?' appears at the bottom right corner. The scene then transitions to show the same person now holding an iPod Nano o, which is white with a black screen. The text 'this is the new' appears at the bottom center of the frame. Finally, the person holds up a pen, indicating they are about to write something.


I have already raised an issue in the transformers repo, and I suspect the problem might originate from the Qwen2LM code. huggingface/transformers#33615

**Comment by liuyijiang1994** [comment]
same， what a annoying problem

**Comment by ZhangYuanhan-AI** [comment]
Try to use bf16 to load the model, instead of fp16.

**Comment by ZhangYuanhan-AI** [comment]
e.g:

tokenizer, model, image_processor, max_length = load_pretrained_model(pretrained, None, model_name, torch_dtype="bfloat16", device_map=device_map)  # Add any other thing you want to pass in llava_model_args

**Comment by WellDonePF** [comment]
> e.g:
> 
> tokenizer, model, image_processor, max_length = load_pretrained_model(pretrained, None, model_name, torch_dtype="bfloat16", device_map=device_map) # Add any other thing you want to pass in llava_model_args

thanks for your reply. it works for easy task, but the problem still exists when task become complex.

### EASY TASK
[QUESTION]
'Deliver concise, accurate answers to questions prompted by the images\n\nContext: <image> <image>\n\nQuestion: What is the special magic Thimphu number in the images?\nRespond with the answer and your thought process for reasoning.'

[MODEL_OUTPUT]
'!425900'

[EXPECTED ANSWER]
'425900'


### COMPLEX TASK
[QUESTION]
I will give you some pictures, and each group of pictures will correspond to a question. Please answer it briefly. You must choose your answer from the Choice List. \n<image><image>Question: what is the title for the name robert kauk ?\nChoice list: \nA. VICE PRESIDENT, BRUSH BEET GROWERS ASS'N\nB. TREASURER, BRUSH BEET GROWERS ASS'N\nC. PRESIDENT, BRUSH BEET GROWERS ASS'N\nD. SECRETARY, BRUSH BEET GROWERS ASS'N\n\nRespond with the answer and your thought process for reasoning."

[MODEL_OUTPUT]
'!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!'

**Comment by Miraclys** [comment]
Could bfloat16 version handle complex scenario? It output nothing e.g. "" when the questions is complex.
